### PR TITLE
Feat/alert/form-validation : Integrate from validation library into alert screens

### PR DESCRIPTION
**Pull request by charliemangano**
# Integrate from validation library into alert screens

## Description
This PR integrates a new library [FromBuilder](https://github.com/jkuatdsc/form-builder) to do form validation in our alert screens. This library is made to work with Jetpack Compose. This tremendously cleans up our screens since the validation logic is handled elsewhere, and better separates the duties MVVM-wise.
This PR closes issue #266 and part of #261.

## Changes
**Explanation of the library:** see [in previous PR](https://github.com/PeriodPals/periodpals/pull/263)

**Actual changes:**
I set up the validation logic in the `AlertViewModel`: the `FormState` is accessible as an attribute of the class and is associated with the corresponding `TextFieldState`s and `Validator`s.
In the screens, I replaced the existing states with the `formState` and offshoots `nameState`, `dobState` etc. I also replaced the validation logic with calls to `validate` and the `onChange` logic with calls to `change`. The error messages are also better handled using the `errorMessage` function.

Seeing the feedback from M2, I added a check for the length of the inputs.

#### Added
None
#### Modified
- `app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt`: implemented the validation logic
- `app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt`, `app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt` and `app/src/main/java/com/android/periodpals/ui/components/AlertComponents.kt`: integrated the library into the screens
- `app/src/test/java/com/android/periodpals/model/alert/AlertViewModelTest.kt`, `app/src/test/java/com/android/periodpals/ui/alert/CreateAlertTest.kt` and `app/src/test/java/com/android/periodpals/ui/alert/EditAlertTest.kt`: update tests

#### Removed
None

## Dependencies Added
None

## Testing
I updated the tests to take into account this new library: in `AlertViewModelTest`, I implemented new checks that verify the `FormState` attribute values, and in `CreateAlertTest` and `EditAlertTest`, I provided the `FormState` as part of the `UserViewModel` mock.

### CODE REVIEW

File: app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt
Comment (francelu) [inline app/src/main/java/com/android/periodpals/model/alert/AlertViewModel.kt:1]: Very good ! Implementation is clear and simple :)

File: app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt [hunk 1]
```diff
@@ -156,53 +159,57 @@ fun CreateAlertScreen(
...
+                    product = stringToProduct(productState.value)!!,
+                    urgency = stringToUrgency(urgencyState.value)!!,
+                    location = locationState.value,
+                    message = messageState.value,
+                    status = Status.CREATED,
+                )
+            alertViewModel.createAlert(
+                alert,
+                onSuccess = { Log.d(TAG, "Alert created") },
+                onFailure = { e -> Log.e(TAG, "createAlert: fail to create alert: ${e.message}") },
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt:209]: **Nitpick**: Is it really useful to state the callbacks ?

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt [hunk 2]
```diff
@@ -176,40 +185,50 @@ fun EditAlertScreen(
...
+                      location = locationState.value,
+                      message = messageState.value,
+                      status = alert.status, // TODO: handle this properly
+                  )
+              alertViewModel.updateAlert(
+                  newAlert,
+                  onSuccess = { Log.d(TAG, "Alert successfully updated") },
+                  onFailure = { e ->
+                    Log.e(TAG, "updateAlert: fail to update alert: ${e.message}")
+                  },
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt:225]: Same nitpick question here :)

File: app/src/test/java/com/android/periodpals/model/alert/AlertViewModelTest.kt
Comment (francelu) [inline app/src/test/java/com/android/periodpals/model/alert/AlertViewModelTest.kt:1]: Good implementation of tests !

**Review by francelu — APPROVED** [review]
## Summary
This pull request adds form validation for the `AlertViewModel` and updates the `CreateAlert` and `EditAlert` screens to utilize this validation. 

LGTM ! Good job as usual.

## Important

#### Code Quality
- The code is generally well-organized and adheres to Kotlin coding standards.
- The new constants for error messages and validation logic improve readability and **maintainability**.
- The use of `companion object` for state names is consistent and follows best practices.

#### Functionality
- The validation logic for product, urgency, location, and message fields ensures that invalid inputs are properly handled.
- The `FormState` and `TextFieldState` integration in `AlertViewModel` is correctly implemented.
- The transformation of location strings to `Location` objects is handled properly.
- Edge cases such as empty and overly long inputs are addressed through validators.

#### Performance
- There are no significant performance concerns introduced by these changes.

#### Security
- Input validation enhances security by preventing invalid or malicious data from being processed.

## Testing
- New tests have been added to `AlertViewModelTest` to verify the validation logic.
- Tests for the `CreateAlertScreen` and `EditAlertScreen` have been updated to reflect the new validation mechanism.

## Documentation
- Inline comments are sufficient and the code is self-explanatory for the most part.

## Suggestions
- Ensure that all potential edge cases are covered in the tests.

## Steps before PR approved
- Resolve comments.

File: app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt (hunk 1 above)
Comment (charliemangano) [reply app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt:209]: The callbacks are not my doing, it looks like a modification because GitHub is not precise enough with the changes: since I removed the `else` bloc by adding a `return` in the `if` bloc, everything that was inside indented one tab less. I won't touch it since it is not in the scope of my task.

To comment on the usefulness of it however, I don't personally think they are explicitly necessary since we have default values, but they can add precision and granularity if needed when debugging.

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt (hunk 2 above)
Comment (charliemangano) [reply app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt:225]: Same here!

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt [hunk 3]
```diff
@@ -131,25 +134,31 @@ fun EditAlertScreen(
 
       // Product dropdown
       ProductField(
-          product = if (product == null) "" else productToPeriodPalsIcon(product!!).textId,
-          onValueChange = { product = stringToProduct(it) },
+          product =
+              if (productState.value.isEmpty()) productState.value
+              else productToPeriodPalsIcon(stringToProduct(productState.value)!!).textId,
```
Comment (agonzalez-r) [inline app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt:139]: I'm pretty sure that here you don't need to do any parsing, you can directly write 
```
product = productState.value
```
I tried it replacing it myself and it seemd to work just fine.

**Review by agonzalez-r — APPROVED** [review]
## Summary
Thank you for your work! It works very efficiently and the code is much better this way. 
## Important (which includes)
#### Code Quality
- Code is readable, clean and consistent. 
#### Functionality
- Nitpick about removing useless parsing in `EditAlert`. 
- Nitpick about extracting the validators to avoid declaring them multiple times (in tests mostly).
#### Performance 
Perfect. 

## Testing
- Thorough tests for the added function in `AlertComponents`, and effectively adapted the other tests. 
- All tests pass. 
## Documentation
- Just one missing little documentation for `fun capitalized()`, but could be omitted due to the very simple nature of this function
## Suggestions
Consider the improvements and nitpicks I left in the comments :)
But overall LGTM

File: app/src/test/java/com/android/periodpals/ui/alert/CreateAlertScreenTest.kt [hunk 4]
```diff
@@ -69,6 +78,41 @@ class CreateAlertScreenTest {
...
+    private const val ERROR_INVALID_PRODUCT = "Please select a product"
+    private const val ERROR_INVALID_URGENCY = "Please select an urgency level"
+    private const val ERROR_INVALID_LOCATION = "Please select a location"
+    private const val ERROR_INVALID_MESSAGE = "Please write your message"
+    private const val ERROR_LOCATION_TOO_LONG =
+        "Location must be less than $MAX_LOCATION_LENGTH characters"
+    private const val ERROR_MESSAGE_TOO_LONG =
+        "Message must be less than $MAX_MESSAGE_LENGTH characters"
+
+    private val productValidators =
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/test/java/com/android/periodpals/ui/alert/CreateAlertScreenTest.kt:94]: Yes, it bothers me as well. The usual way we would do this is by mocking everything, but I already tried on my previous form validation PRs and couldn't do it. 

The one thing I think would be viable is to make them `public` to be able to access them from the tests (which essentially is the same as extracting them in a separate file). However that would break encapsulation and I didn't want to do that without collective agreement from everyone. We can talk about it more after the meeting to see if we find a solution

**Comment by charliemangano** [comment]
Following a bug for GitHub, @agonzalez-r 's review is not visible anymore. Here are screenshots of it:

image
image
